src/api/fossa.rs

```diff
@@ -84,3 +210,216 @@ impl TryFrom<String> for Key {
+                Transport::Http { endpoint, .. } => endpoint.to_string(),
+            },
+        };
+
+        match reference {
+            Reference::Git(reference) => match reference {
+                git::Reference::Branch { name: branch, head } => Self {
+                    name,
+                    branch: Some(branch.to_string()),
+                    revision: head.to_string(),
+                },
+                git::Reference::Tag { name: tag, .. } => Self {
+                    name,
+                    branch: None,
+                    revision: tag.to_string(),
+                },
+            },
+        }
+    }
+}
+
+impl Display for ProjectMetadata {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let name = &self.name;
+        let revision = &self.revision;
+        match &self.branch {
+            Some(branch) => write!(f, "{name}@{revision} ({branch})"),
+            None => write!(f, "{name}@{revision}"),
+        }
+    }
+}
+
+/// Metadata from FOSSA CLI.
+#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize, new)]
+pub struct CliMetadata {
+    version: Version,
+}
+
+/// Upload the scan results for a project.
+///
+/// In the future we'd like to have this method be made available via trait so we can test.
+/// I ran out of time this time around though.
+#[tracing::instrument(skip(source_units))]
+pub async fn upload_scan(
+    opts: &Config,
+    project: &ProjectMetadata,
+    cli: &CliMetadata,
+    source_units: SourceUnits,
+) -> Result<Locator, Error> {
+    let url = opts.endpoint().join("api/builds/custom")?;
+
+    let locator = Locator::builder()
+        .fetcher(Fetcher::Custom)
+        .project(&project.name)
+        .revision(&project.revision)
+        .build();
+    let package_locator = locator.clone().into_package();
+
+    // We don't currently include metadata such as team/policies/etc.
+    // We can add it to integration config as users request it though.
+    let mut query = vec![
+        ("title", package_locator.to_string()),
+        ("locator", locator.to_string()),
+        ("cliVersion", cli.version.to_string()),
+        ("managedBuild", String::from("true")),
+    ];
+    if let Some(branch) = &project.branch {
+        query.push(("branch", branch.to_string()));
+    }
+
+    let req = new_client()?
+        .post(url)
+        .bearer_auth(opts.key().expose_secret())
+        .query(&query)
+        .header(CONTENT_TYPE, "application/json")
+        .body(source_units.to_string());
+
+    run_request::<UploadResponse>(req)
+        .await
+        .change_context_lazy(|| Error::upload_scan(&locator, &source_units))?
+        .into()
+}
+
+impl Endpoint {
+    /// Make a GET request against the FOSSA server with the provided route,
+    /// which is joined to the base.
+    #[tracing::instrument]
+    async fn get<T: DeserializeOwned>(&self, route: &str, token: &Key) -> Result<T, Error> {
```

Comment:
I think should to refactor `broker fix` to use this.

Reply:
Although, if `lookup` is never used, is this ever used?

Reply:
It's not, it was earlier until I realized we don't need to do this. But the code is written, so...